core/src/main/java/org/apache/struts2/interceptor/FetchMetadataInterceptor.java

```diff
@@ -56,32 +58,33 @@
 
     private final Set<String> exemptedPaths = new HashSet<String>();
     private final ResourceIsolationPolicy resourceIsolationPolicy = new DefaultResourceIsolationPolicy();
+    static final String VARY_HEADER_VALUE = String.format("%s,%s,%s", DefaultResourceIsolationPolicy.SEC_FETCH_DEST_HEADER, DefaultResourceIsolationPolicy.SEC_FETCH_SITE_HEADER, DefaultResourceIsolationPolicy.SEC_FETCH_MODE_HEADER);
 
-    public FetchMetadataInterceptor(Set<String> exemptedPaths){
-        this.exemptedPaths.addAll(exemptedPaths);
+    public void setExemptedPaths(String paths){
+        this.exemptedPaths.addAll(TextParseUtil.commaDelimitedStringToSet(paths));
     }
 
     @Override
     public void beforeResult(ActionInvocation invocation, String resultCode) {
         // Add Vary headers
         HttpServletResponse response = invocation.getInvocationContext().getServletResponse();
-        response.setHeader(resourceIsolationPolicy.VARY_HEADER, resourceIsolationPolicy.SEC_FETCH_DEST_HEADER + ", "
-                + resourceIsolationPolicy.SEC_FETCH_SITE_HEADER + ", " + resourceIsolationPolicy.SEC_FETCH_MODE_HEADER);
+        response.setHeader(DefaultResourceIsolationPolicy.VARY_HEADER, VARY_HEADER_VALUE);
     }
 
     @Override
     public String intercept(ActionInvocation invocation) throws Exception {
         ActionContext context = invocation.getInvocationContext();
         HttpServletRequest request = context.getServletRequest();
 
+        // Adds listener that operates between interceptor and result rendering to set Vary headers
+        invocation.addPreResultListener(this);
+
         // Apply exemptions: paths/endpoints meant to be served cross-origin
-        if (!(resourceIsolationPolicy.isRequestAllowed(request) || this.exemptedPaths.contains(request.getContextPath()))) {
+        if (!resourceIsolationPolicy.isRequestAllowed(request) && !this.exemptedPaths.contains(request.getContextPath())) {
```

Comment:
Let's flip these two conditions so that the policy is not called if the path is exempted or split into two ifs. Please see previous comment on this line about Java short circuit evaluation. Alternatively, use the following suggestion.
```suggestion
        if (exemptedPaths.contains(request.getContextPath()) {
            return invocation.invoke();
        }
        if (resourceIsolationPolicy.isRequestAllowed(request)) {
            return invocation.invoke();
        }
        
        beforeResult(invocation, String.valueOf(HttpServletResponse.SC_FORBIDDEN));
        return String.valueOf(HttpServletResponse.SC_FORBIDDEN);
```

Reply:
Misread previous comment on this part :( thanks!

Reply:
In general, huge fan of being able to bail early from codepaths that run on every request